lightningd/opening_control.c

```diff
@@ -358,6 +358,210 @@ static void opening_funder_start_replied(struct subd *openingd, const u8 *resp,
+		script = tal_arr(utx->outputs, u8, out.script_len);
+		memcpy(script, out.script, out.script_len);
+
+		utx->outputs[i] = new_tx_output(utx->outputs,
+						amount, script);
+	}
+
+	/* Update transaction fields on utx */
+	utx->tx = tal_free(utx->tx);
+	utx->tx = tal_steal(utx, tx);
+	bitcoin_txid(tx, &utx->txid);
+
+	return true;
+}
+#endif /* EXPERIMENTAL_FEATURES */
+
+static void opening_opener_finished(struct subd *openingd, const u8 *resp,
+				    const int *fds,
+				    struct uncommitted_channel *uc)
+{
+#if EXPERIMENTAL_FEATURES
+	struct channel_info channel_info;
+	struct bitcoin_txid funding_txid;
+	u16 funding_txout;
+	struct bitcoin_signature remote_commit_sig;
+	struct bitcoin_tx *remote_commit, *funding_tx;
+	u32 feerate, feerate_funding;
+	struct amount_sat local_funding, total_funding;
+	struct channel *channel;
+	struct lightningd *ld = openingd->ld;
+	u8 *remote_upfront_shutdown_script;
+	struct per_peer_state *pps;
+	struct bitcoin_tx_input **remote_ins;
+	struct funding_channel *fc = uc->fc;
+	struct unreleased_tx *utx = fc->utx;
+	u32 *input_map;
+	struct witness_stack **witnesses = NULL;
+
+	/* This is a new channel_info.their_config so set its ID to 0 */
+	channel_info.their_config.id = 0;
+
+	if (!fromwire_opening_dual_funding_signed(resp, resp,
+						  &pps,
+						  &remote_commit,
+						  &remote_commit_sig,
+						  &funding_tx,
+						  &funding_txout,
+						  &remote_ins,
+						  &input_map,
+						  &local_funding,
+						  &channel_info.their_config,
+						  &channel_info.theirbase.revocation,
+						  &channel_info.theirbase.payment,
+						  &channel_info.theirbase.htlc,
+						  &channel_info.theirbase.delayed_payment,
+						  &channel_info.remote_per_commit,
+						  &channel_info.remote_fundingkey,
+						  &feerate,
+						  &feerate_funding,
+						  &uc->our_config.channel_reserve,
+						  &remote_upfront_shutdown_script)) {
+		log_broken(fc->uc->log,
+			   "bad OPENING_DUAL_FUNDING_SIGNED %s",
+			   tal_hex(resp, resp));
+		was_pending(command_fail(fc->cmd, LIGHTNINGD,
+					 "bad OPENING_DUAL_FUNDING_SIGNED %s",
+					 tal_hex(fc->cmd, resp)));
+		goto cleanup;
+	}
+
+	remote_commit->chainparams = get_chainparams(openingd->ld);
+	per_peer_state_set_fds_arr(pps, fds);
+
+	/* openingd should never accept them funding channel in this case. */
+	if (peer_active_channel(uc->peer)) {
+		log_broken(uc->log, "openingd accepted peer funding channel");
+		uncommitted_channel_disconnect(uc, "already have active channel");
+		goto cleanup;
+	}
+
+	/* Update the funding tx that we have for this */
+	if (!update_unreleased_tx(utx, funding_tx, input_map, witnesses)) {
+		log_broken(uc->log, "Unable to update unreleased tx");
+		uncommitted_channel_disconnect(uc, "internal error in creating updated tx");
+		goto cleanup;
+	}
+
+
+	/* Pull out the funding amount */
+	total_funding.satoshis  /* Raw: type conversion */
+		= utx->tx->wtx->outputs[funding_txout].satoshi;  /* Raw: type conversion */
```

Comment:
Aside: I wonder if we should have a tx_output_amount(utx->tx, funding_txout) function which does this?